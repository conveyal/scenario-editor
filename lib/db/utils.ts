import mapValues from 'lodash/mapValues'
import {ObjectID} from 'mongodb'

export function serializeCollectionResults(obj: unknown) {
  if (obj instanceof Date) return obj.toISOString()
  if (obj instanceof ObjectID) return obj.toString()
  if (Array.isArray(obj)) return obj.map(serializeCollectionResults)
  if (typeof obj === 'object') return mapValues(obj, serializeCollectionResults)
  return obj
}
