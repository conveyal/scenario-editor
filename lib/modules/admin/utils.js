import {connect} from 'react-redux'
import {createStructuredSelector} from 'reselect'

import * as actions from './actions'
import * as select from './selectors'

export function fullyConnect(Component) {
  return connect(createStructuredSelector(select), actions)(Component)
}

const add0 = n => (n >= 10 ? n : `0${n}`)

export function msToDuration(ms) {
  const durationS = Math.floor((Date.now() - ms) / 1000)
  const s = durationS % 60
  const m = Math.floor((durationS / 60) % 60)
  const h = Math.floor((durationS / 60 / 60) % 60)

  return `${add0(h)}:${add0(m)}:${add0(s)}`
}

/**
 * Example commit string: "v3.4.0-118-gbe7199c". Take the last section after
 * splitting on the hyphens and remove the "g".
 */
const githubUrl = 'https://github.com/conveyal/r5/commits'
export function createR5Url(workerCommit) {
  // not a release version
  if (workerCommit.indexOf('-') !== -1) {
    const hash = workerCommit.split('-')[2].slice(1) //
    return `${githubUrl}/${hash}`
  }
  return `${githubUrl}/${workerCommit}`
}

export function createWorkerUrl(instanceId, region) {
  return `https://${region}.console.aws.amazon.com/ec2/v2/home?region=${region}#Instances:instanceId=${instanceId};sort=tag:Name`
}

const C1 = '#2389c9'
const C2 = '#36c923'
const C3 = '#aab4bb'

/**
 * Create a progress bar with a linear-gradient.
 */
export function createLinearGradientForJob(c, d, t) {
  const p1 = (c / t) * 100
  const p2 = (d / t) * 100
  return `linear-gradient(to right,
    ${C1} 0%,
    ${C1} ${p1}%,
    ${C2} ${p1}%,
    ${C2} ${p2}%,
    ${C3} ${p2}%,
    ${C3} 100%)`
}
