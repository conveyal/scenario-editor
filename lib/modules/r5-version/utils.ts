import {
  MINIMUM_R5_VERSION,
  VERSION_PARSE_REGEX,
  OLD_VERSION_PARSE_REGEX
} from './constants'

/**
 * Convert an r5 version to a number for comparing
 * NB Assumes no major, minor, patch, or commit version exceeds 100
 */
export function versionToNumber(version?: string): number {
  const versionString = version || ''
  if (versionString.match(/\./g) && versionString.match(/\./g).length == 1) {
    const versionMatch = versionString.match(VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [, major = '0', minor = '0', commit = '0'] = versionMatch

      return (
        parseInt(major) * 10000 + parseInt(minor) * 100 + parseInt(commit) / 100
      )
    }

    return 0
  } else {
    const versionMatch = versionString.match(OLD_VERSION_PARSE_REGEX)
    if (versionMatch) {
      const [
        ,
        major = '0',
        minor = '0',
        patch = '0',
        commit = '0'
      ] = versionMatch

      return (
        parseInt(major) * 10000 +
        parseInt(minor) * 100 +
        parseInt(patch) +
        parseInt(commit) / 100
      )
    }

    return 0
  }
}

/**
 * Check if the r5 version is older than a given minimum.
 */
export function workerVersionInRange(
  version: string,
  min: string = MINIMUM_R5_VERSION,
  max = 'v1000.0.0'
): boolean {
  const versionNumber = versionToNumber(version)
  return (
    versionNumber >= versionToNumber(min) &&
    versionNumber <= versionToNumber(max)
  )
}

/**
 * Check if the r5 version is a patch version or in a given range
 */
export function workerVersionTestOrInRange(
  version: string,
  min: string,
  max?: string
): boolean {
  const isTestVersion = (version || '').search(/-/g) !== -1
  return isTestVersion || workerVersionInRange(version, min, max)
}
