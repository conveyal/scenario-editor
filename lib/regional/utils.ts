function getDefaultCutoff(a: CL.RegionalAnalysis): number {
  if (Array.isArray(a.cutoffsMinutes))
    return a.cutoffsMinutes[Math.floor(a.cutoffsMinutes.length / 2)]
  return a.cutoffMinutes
}

function getDefaultDestinationPointSet(a: CL.RegionalAnalysis): string | void {
  const ids = a.destinationPointSetIds
  if (Array.isArray(ids)) return ids[0]
}

function getDefaultPercentile(a: CL.RegionalAnalysis): number {
  if (Array.isArray(a.travelTimePercentiles))
    return a.travelTimePercentiles[
      Math.floor(a.travelTimePercentiles.length / 2)
    ]
  return a.travelTimePercentile
}

/**
 * Choose
 */
export function getDefaultVariants(analysis: CL.RegionalAnalysis) {
  return {
    cutoff: getDefaultCutoff(analysis),
    destinationPointSetId: getDefaultDestinationPointSet(analysis),
    percentile: getDefaultPercentile(analysis)
  }
}

export function getComparisonVariants(comparison: CL.RegionalAnalysis) {
  return {
    comparisonCutoff: getDefaultCutoff(comparison),
    comparisonDestinationPointSetId: getDefaultDestinationPointSet(comparison),
    comparisonPercentile: getDefaultPercentile(comparison)
  }
}

/**
 * Don't compare incomparable analyses.
 */
export function getValidComparisonAnalyses(
  allAnalyses: CL.RegionalAnalysis[],
  activeJobs: CL.RegionalJob[],
  analysis: CL.RegionalAnalysis
) {
  return allAnalyses.filter(
    (c) =>
      activeJobs.findIndex((job) => job.jobId === c._id) === -1 &&
      c.request.originPointSetKey == null &&
      c.zoom === analysis.zoom &&
      c.width === analysis.width &&
      c.height === analysis.height &&
      c.north === analysis.north &&
      c.west === analysis.west
  )
}
