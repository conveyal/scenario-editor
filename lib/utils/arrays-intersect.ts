// Do the arrays intersect?
export default function arraysIntersect<T>(a1: T[], a2: T[]) {
  for (const v of a1) {
    if (a2.indexOf(v) > -1) return true
  }
  return false
}
