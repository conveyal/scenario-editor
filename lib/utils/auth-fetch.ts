import {parse} from 'cookie'

import {safeFetch, SafeResponse} from './safe-fetch'

export function createAuthHeaders(user?: CL.User) {
  const headers = {
    Authorization: `bearer ${user?.idToken}`
  }

  // Add the admin access group for administrators if it exists.
  if (user?.accessGroup === process.env.NEXT_PUBLIC_ADMIN_ACCESS_GROUP) {
    const adminTempAccessGroup = parse(document.cookie).adminTempAccessGroup
    if (adminTempAccessGroup?.length > 0) {
      headers['X-Conveyal-Access-Group'] = adminTempAccessGroup
    }
  }
  return headers
}

/**
 * Fetch wrapper that includes authentication. Defaults headers to JSON.
 */
export default function authFetch<T>(
  url: string,
  user: CL.User,
  parser: (res: Response) => Promise<T> = (res) => res.json(),
  options?: RequestInit
): Promise<SafeResponse<T>> {
  return safeFetch(url, parser, {
    mode: 'cors',
    ...options,
    headers: {
      ...createAuthHeaders(user),
      ...options?.headers
    }
  })
}
