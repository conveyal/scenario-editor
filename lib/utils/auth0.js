// @flow
import message from '@conveyal/woonerf/message'
import Auth0Lock from 'auth0-lock'

const CLIENT_ID = process.env.AUTH0_CLIENT_ID
const DOMAIN = process.env.AUTH0_DOMAIN

export const authIsRequired = CLIENT_ID && DOMAIN

const localStorage = window && window.localStorage
  ? window.localStorage
  : {
    getItem () {},
    setItem () {}
  }

const lock = authIsRequired
  ? new Auth0Lock(CLIENT_ID, DOMAIN, {
    auth: {
      params: {
        scope: 'openid email profile analyst offline_access'
      },
      redirect: false
    },
    allowSignUp: false,
    theme: {
      logo: 'https://s3-eu-west-1.amazonaws.com/analyst-logos/conveyal-128x128.png',
      primaryColor: '#2389c9'
    },
    closable: false,
    languageDictionary: {
      title: message('authentication.logIn')
    }
  })
  : {
    checkSession () {},
    getUserInfo () {},
    hide () {},
    logout () {},
    on () {},
    show () {}
  }

function flashError (error) {
  console.error(error)
  lock.show({
    flashMessage: {
      type: 'error',
      text: error.error_description
    }
  })
}

// Set the default handler for auth errors
lock.on('authorization_error', flashError)
lock.on('unrecoverable_error', flashError)

export function login (next: (any) => void) {
  function handleAuthResult (authResult) {
    console.log('handleAuthResult', authResult)
    lock.getUserInfo(authResult.accessToken, (error, profile) => {
      if (error) {
        localStorage.removeItem('user')
        lock.show({
          flashMessage: {
            type: 'error',
            text: error.error_description
          }
        })
      } else {
        localStorage.setItem('user', JSON.stringify(profile))
        next({
          ...authResult,
          // idToken: authResult.accessToken, // for woonerf/fetch
          profile
        })
        lock.hide()
      }
    })
  }

  lock.checkSession({}, (error, authResult) => {
    console.log('checkSession', error, authResult)
    if (error || !authResult) {
      lock.show()
      lock.on('authenticated', handleAuthResult)
    } else {
      try {
        const profile = JSON.parse(localStorage.getItem('user'))
        next({
          ...authResult,
          // idToken: authResult.accessToken, // for woonerf/fetch
          profile
        })
        lock.hide()
      } catch (e) {
        console.error('error processing user profile from localStorage', e)
        handleAuthResult(authResult)
      }
    }
  })
}

export function logout () {
  localStorage.removeItem('user')
  lock.logout({returnTo: `${window.location.origin}/login`})
}
