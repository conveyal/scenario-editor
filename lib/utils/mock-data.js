import multi from '@conveyal/woonerf/store/multi'
import promise from '@conveyal/woonerf/store/promise'
import {middleware as fetchMiddleware} from '@conveyal/woonerf/fetch'
import configureStore from 'redux-mock-store'
import thunkMiddleware from 'redux-thunk'
import {TextEncoder} from 'text-encoding'

import {responseToSurface} from '../actions/analysis'
import {ADD_TRIP_PATTERN} from '../constants'
import {TRAVEL_TIME_PERCENTILES} from '../constants/analysis'

export const makeMockStore = configureStore([fetchMiddleware, multi, promise, thunkMiddleware])

export const mockSegmentGeom = {
  type: 'LineString',
  coordinates: [
    [
      -122.0246,
      36.9707
    ],
    [
      -122.0279,
      37.049
    ],
    [
      -121.9799,
      37.2299
    ],
    [
      -121.9445,
      37.324
    ],
    [
      -121.936,
      37.353
    ],
    [
      -121.924,
      37.365
    ]
  ]
}

export const mockSegment = {
  type: 'Feature',
  properties: {},
  geometry: mockSegmentGeom
}

const mockStop1 = {
  stop_id: '1',
  stop_lat: 36.9707,
  stop_lon: -122.0246,
  stop_name: 'Santa Cruz'
}

const mockStop2 = {
  stop_id: '2',
  stop_lat: 37.049,
  stop_lon: -122.0279,
  stop_name: 'Scotts Valley'
}

export const mockStops = [mockStop1, mockStop2]

export const mockPattern = {
  geometry: mockSegmentGeom,
  stops: mockStops,
  trips: [
    {
      trip_id: 'abcd',
      start_time: 12345,
      trip_short_name: 'The Express',
      trip_headsign: 'To Downtown',
      duration: 1234
    }
  ]
}

export const mockRoute = {
  id: 'route1',
  patterns: [mockPattern]
}

export const mockFeed = {
  id: '1',
  routes: [mockRoute],
  routesById: {
    route1: mockRoute
  },
  stopsById: {
    '1': mockStop1,
    '2': mockStop2
  },
  stops: [mockStop1, mockStop2]
}

export const mockMapState = {
  allowExtend: true,
  extendFromEnd: true,
  followRoad: true,
  state: 'add-trip-pattern',
  modificationId: '1234'
}

export const mockTimetable = {
  name: 'Test timetable',
  segmentSpeeds: [],
  dwellTime: 10,
  monday: true,
  tuesday: true,
  wednesday: true,
  thursday: true,
  friday: true,
  saturday: false,
  sunday: false,
  headwaySecs: 900, // 15 minutes
  startTime: 28800, // 8am
  endTime: 57600, // 4pm
  patternTrips: ['abcd'],
  sourceTrip: 'abcd'
}

export const mockModification = {
  bidirectional: false,
  entries: [],
  feed: mockFeed.id,
  id: '1234',
  name: 'Test Modification',
  routes: ['route1'],
  segments: [],
  showOnMap: false,
  timetables: [mockTimetable],
  trips: ['abcd'],
  type: ADD_TRIP_PATTERN,
  variants: [0]
}

export const mockProject = {
  bounds: {
    north: 39,
    east: -76,
    south: 38,
    west: -77
  },
  description: 'Project description',
  id: '1',
  name: 'Mock Project'
}

export const mockScenario = {
  bundleId: '1',
  feeds: [mockFeed],
  feedsById: {
    [mockFeed.id]: mockFeed
  },
  id: '1',
  name: 'Mock Scenario',
  projectId: '1',
  variants: ['Default']
}

export const mockStores = {
  init: {
    analysis: {
      isFetchingIsochrone: false,
      isochroneCutoff: 60
    },
    mapState: {
      activeTrips: [],
      center: {
        lat: 38.8886,
        lon: -77.043
      },
      components: [],
      zoom: 12
    },
    network: {
      error: null,
      outstandingRequests: 0
    },
    project: {
      currentProject: null,
      projects: [],
      projectsById: {}
    },
    routing: {
      locationBeforeTransitions: null
    },
    scenario: {
      bundles: [],
      bundlesById: {},
      currentBundle: {
        id: '1'
      },
      currentScenario: mockScenario,
      feeds: [mockFeed],
      feedsById: {
        [mockFeed.id]: mockFeed
      },
      modifications: [mockModification],
      modificationsById: {
        [mockModification.id]: mockModification
      },
      modificationsByType: {
        [mockModification.type]: [mockModification]
      },
      scenarios: [],
      scenariosById: {},
      variants: ['Default']
    },
    user: {
      idToken: null
    }
  }
}

export const mockScenarioApplicationError = {
  'modificationId': 'Add Trip Pattern',
  'title': "Error while applying the modification entitled 'Add Trip Pattern'.",
  'messages': [
    'Headway is not greater than zero.'
  ]
}

export const mockProfileRequest = {
  date: '2016-01-16',
  fromTime: 25200,
  toTime: 32400,
  accessModes: 'WALK',
  directModes: 'WALK',
  egressModes: 'WALK',
  transitModes: 'TRANSIT',
  walkSpeed: 1.3888888888888888,
  bikeSpeed: 4.166666666666667,
  carSpeed: 20,
  streetTime: 90,
  maxWalkTime: 20,
  maxBikeTime: 20,
  maxCarTime: 45,
  minBikeTime: 10,
  minCarTime: 10,
  suboptimalMinutes: 5,
  reachabilityThreshold: 0,
  bikeSafe: 1,
  bikeSlope: 1,
  bikeTime: 1,
  bikeTrafficStress: 4,
  monteCarloDraws: 200,
  maxRides: 4
}

export const mockRegionalAnalyses = [
  {
    zoom: 9,
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 1',
    grid: 'Jobs_total',
    scenarioId: '1234',
    variant: 0,
    bundleId: '5678',
    cutoffMinutes: 60,
    id: 'abcd',
    request: mockProfileRequest,
    // percentile -1 indicates this analysis was done using Andrew Owen-style average instantaneous accessibility
    travelTimePercentile: -1
  },
  {
    zoom: 9,
    width: 300,
    height: 300,
    north: 20024,
    west: 20015,
    name: 'ANALYSIS 2',
    grid: 'Jobs_total',
    scenarioId: '4321',
    variant: 1,
    bundleId: '5678',
    cutoffMinutes: 60,
    id: 'efgh',
    request: mockProfileRequest,
    travelTimePercentile: 42
  }
]

export const mockGrid = {
  zoom: 9,
  width: 300,
  height: 300,
  north: 20024,
  west: 20015,
  data: [],
  min: 0,
  max: 2000
}

export const mockBookmark = {
  name: 'Bookmark',
  profileRequest: mockProfileRequest,
  isochroneCutoff: 42,
  destinationGrid: 'Workers_total',
  projectId: 'PROJECT'
}

export const mockSurfaceParameters = {
  width: 5,
  height: 5,
  nSamples: 5,
  zoom: 9,
  north: 40000,
  west: 40000
}

// this is a function because responseToSurface is destructive and can lead to test leakage if we
// don't have a brand-new surface each time
export const createMockSurfaceData = () => {
  const {width, height, nSamples, zoom, north, west} = mockSurfaceParameters

  const metadata = {
    scenarioApplicationWarnings: [mockScenarioApplicationError]
  }

  const metadataEncoded = new TextEncoder().encode(JSON.stringify(metadata))
  const bufSize = 9 * 4 + // header
    width * height * nSamples * 4 + // data
    metadataEncoded.length // metadata

  const buffer = new ArrayBuffer(bufSize)
  const surfaceData = new Int32Array(buffer, 0, 9 + width * height * nSamples)

  // text header ("ACCESSGR")
  surfaceData[0] = 1162036033 // ACCE
  surfaceData[1] = 1380406099 // SSGR

  // version
  surfaceData[2] = 0

  surfaceData[3] = zoom
  surfaceData[4] = west
  surfaceData[5] = north
  surfaceData[6] = width
  surfaceData[7] = height
  surfaceData[8] = nSamples

  // phew, header is done
  for (let y = 0, pixel = 9; y < height; y++) {
    for (let x = 0; x < width; x++) {
      let prev = 0
      for (let z = 0; z < nSamples; z++, pixel++) {
        const pixelVal = getMockSurfaceValue(x, y, z)
        surfaceData[pixel] = pixelVal - prev // values are delta-coded within each pixel
        prev = pixelVal
      }
    }
  }

  // copy metadata
  const metadataBuffer = new Uint8Array(buffer, buffer.byteLength - metadataEncoded.byteLength, metadataEncoded.length)
  for (let i = 0; i < metadataEncoded.length; i++) {
    metadataBuffer[i] = metadataEncoded[i]
  }

  return buffer
}

export const mockSurface = responseToSurface(createMockSurfaceData())

// Return the value that should be stored in sample z of pixel x, y of the mock surface
export function getMockSurfaceValue (x, y, z) {
  // this used to simply return z, 2 * z and 3 * z when we had an array of [5, 10, 15 . . . 95]th
  // percentiles. However, when we switched to requesting specific percentiles that stopped working.
  // This translates the position in the access grid into a percentile, then subtracts five and divides
  // by five to get the position in the old array, so that tests written assuming the old structure continue
  // to pass.
  if (x < 2 && y < 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y >= 2) return (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x >= 2 && y < 2) return 2 * (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
  if (x < 2 && y >= 2) return 3 * (TRAVEL_TIME_PERCENTILES[z] - 5) / 5
}

const destinationGridData = new Int32Array(mockSurfaceParameters.width * mockSurfaceParameters.height)
// top half of grid is 1, bottom half is 2
for (let i = 0; i < 3 * mockSurfaceParameters.width; i++) destinationGridData[i] = 1
for (let i = 3 * mockSurfaceParameters.width; i < destinationGridData.length; i++) destinationGridData[i] = 2

export const mockDestinationGrid = {
  ...mockSurfaceParameters,
  west: mockSurfaceParameters.west - 1, // shift slightly so that we are testing misaligned grids
  north: mockSurfaceParameters.north - 1,
  data: destinationGridData
}

export const releaseVersions = [ 'v3.0.0', 'v2.2.0', 'v2.0.0' ]
export const allVersions = [ 'v3.0.0',
  'v2.2.0-2-g789df3e',
  'v2.2.0',
  'v2.0.0-19-g589f32f',
  'v2.0.0' ]
