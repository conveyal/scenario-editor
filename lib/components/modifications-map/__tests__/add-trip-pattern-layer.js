// @flow
import React from 'react'
import {Map} from 'react-leaflet'

import AddTripPatternLayer from '../add-trip-pattern-layer'
import {mockWithProvider, mockSegment} from '../../../utils/mock-data'

describe('Project-Map > AddTripPatternLayer', () => {
  it('renders correctly', () => {
    // mount component
    expect(mockWithProvider(
      <Map>
        <AddTripPatternLayer segments={[mockSegment]} bidirectional={false} />
      </Map>
    ).snapshot()).toMatchSnapshot()
  })
})
