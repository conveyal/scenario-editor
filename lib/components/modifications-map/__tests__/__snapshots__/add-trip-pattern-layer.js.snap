// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project-Map > AddTripPatternLayer renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Map
    animate={false}
    useFlyTo={false}
  >
    <div>
      <AddTripPatternLayer
        bidirectional={false}
        segments={
          Array [
            {
            "fromStopId": "1",
            "geometry": {
              "type": "LineString",
              "coordinates": [
                [
                  "-122.02460000",
                  "36.97070000"
                ],
                [
                  "-122.02790000",
                  "37.04900000"
                ],
                [
                  "-121.97990000",
                  "37.22990000"
                ],
                [
                  "-121.94450000",
                  "37.32400000"
                ],
                [
                  "-121.93600000",
                  "37.35300000"
                ],
                [
                  "-121.92400000",
                  "37.36500000"
                ]
              ]
            },
            "properties": {},
            "spacing": "3000.00000000",
            "stopAtEnd": false,
            "stopAtStart": false,
            "type": "Feature"
          },
          ]
        }
      >
        <g>
          <Polyline
            color="#2389c9"
            positions={
              Array [
                Object {
                  "lat": 36.9707,
                  "lng": -122.0246,
                },
                Object {
                  "lat": 37.049,
                  "lng": -122.0279,
                },
                Object {
                  "lat": 37.2299,
                  "lng": -121.9799,
                },
                Object {
                  "lat": 37.324,
                  "lng": -121.9445,
                },
                Object {
                  "lat": 37.353,
                  "lng": -121.936,
                },
                Object {
                  "lat": 37.365,
                  "lng": -121.924,
                },
              ]
            }
            weight={3}
          />
          <DirectionalMarkers
            color="#2389c9"
            patterns={
              Array [
                Object {
                  "geometry": Object {
                    "coordinates": Array [
                      Array [
                        -122.0246,
                        36.9707,
                      ],
                      Array [
                        -122.0279,
                        37.049,
                      ],
                      Array [
                        -121.9799,
                        37.2299,
                      ],
                      Array [
                        -121.9445,
                        37.324,
                      ],
                      Array [
                        -121.936,
                        37.353,
                      ],
                      Array [
                        -121.924,
                        37.365,
                      ],
                    ],
                  },
                },
              ]
            }
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 0,
                "index": 0,
                "lat": 36.9707,
                "lon": -122.0246,
                "stopId": "1",
              }
            }
            color="#c92336"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 3000,
                "index": 0,
                "lat": 36.99765592006801,
                "lon": -122.02573607325958,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 6000,
                "index": 0,
                "lat": 37.02461184013602,
                "lon": -122.02687214651915,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 9000,
                "index": 0,
                "lat": 37.05151359890414,
                "lon": -122.02723304175015,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 12000,
                "index": 0,
                "lat": 37.07790094264308,
                "lon": -122.02023142483766,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 15000,
                "index": 0,
                "lat": 37.10428828638203,
                "lon": -122.01322980792517,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 18000,
                "index": 0,
                "lat": 37.13067563012098,
                "lon": -122.00622819101268,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 21000,
                "index": 0,
                "lat": 37.15706297385993,
                "lon": -121.99922657410019,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 24000,
                "index": 0,
                "lat": 37.183450317598876,
                "lon": -121.9922249571877,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 27000,
                "index": 0,
                "lat": 37.20983766133782,
                "lon": -121.98522334027521,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 30000,
                "index": 0,
                "lat": 37.23609341419046,
                "lon": -121.97757006522484,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 33000,
                "index": 0,
                "lat": 37.261931772765294,
                "lon": -121.96784979005429,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 36000,
                "index": 0,
                "lat": 37.28777013134013,
                "lon": -121.95812951488374,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 39000,
                "index": 0,
                "lat": 37.31360848991497,
                "lon": -121.94840923971319,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 42000,
                "index": 0,
                "lat": 37.339703323688006,
                "lon": -121.93989730167766,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
          <CircleMarker
            center={
              Object {
                "autoCreated": true,
                "distanceFromStart": 45000,
                "index": 0,
                "lat": 37.36342593800434,
                "lon": -121.92557406199568,
                "stopId": undefined,
              }
            }
            color="#2389c9"
            fill={true}
            fillColor="#fff"
            fillOpacity={1}
            radius={4.066666666666666}
            weight={1.3555555555555554}
          />
        </g>
      </AddTripPatternLayer>
    </div>
  </Map>
</Provider>
`;
